components/telemetry-operator/internal/validation/plugin_validator.go

```diff
@@ -4,102 +4,119 @@ import (
+func NewPluginValidator(deniedFilterPlugins, deniedOutputPlugins []string) PluginValidator {
 	return &pluginValidator{
-		supportedFilterPlugins: supportedFilerPlugins,
-		supportedOutputPlugins: supportedOutputPlugins,
-		deniedFilterPlugins:    deniedFilterPlugins,
-		deniedOutputPlugins:    deniedOutputPlugins,
+		deniedFilterPlugins: deniedFilterPlugins,
+		deniedOutputPlugins: deniedOutputPlugins,
 	}
 }
 
-func (v *pluginValidator) Validate(logPipeline *telemetryv1alpha1.LogPipeline, logPipelines *telemetryv1alpha1.LogPipelineList) error {
-
+// ContainsCustomPlugin returns true if the pipeline
+// contains any custom filters or outputs
+func (pv *pluginValidator) ContainsCustomPlugin(logPipeline *telemetryv1alpha1.LogPipeline) bool {
 	for _, filterPlugin := range logPipeline.Spec.Filters {
-		if err := checkIfPluginIsValid("filter", filterPlugin.Content, logPipeline, v.supportedFilterPlugins, v.deniedFilterPlugins, logPipelines); err != nil {
-			return err
+		if filterPlugin.Custom != "" {
+			return true
+		}
+	}
+	if logPipeline.Spec.Output.Custom != "" {
+		return true
+	}
+	for _, f := range logPipeline.Spec.Filters {
+		if f.Custom != "" {
+			return true
 		}
 	}
+	return false
+}
 
-	for _, outputPlugin := range logPipeline.Spec.Outputs {
-		if err := checkIfPluginIsValid("output", outputPlugin.Content, logPipeline, v.supportedOutputPlugins, v.deniedOutputPlugins, logPipelines); err != nil {
+// Validate returns an error if validation fails
+// because of using denied plugins or errors in match conditions
+// for filters or outputs.
+func (pv *pluginValidator) Validate(logPipeline *telemetryv1alpha1.LogPipeline, logPipelines *telemetryv1alpha1.LogPipelineList) error {
+	err := pv.validateFilters(logPipeline, logPipelines)
+	if err != nil {
+		return errors.Wrap(err, " error validating Filter plugins")
+	}
+	err = pv.validateOutputs(logPipeline, logPipelines)
+	if err != nil {
+		return errors.Wrap(err, " error validating Output plugins")
+	}
+	return nil
+}
+
+func (pv *pluginValidator) validateFilters(pipeline *telemetryv1alpha1.LogPipeline, pipelines *telemetryv1alpha1.LogPipelineList) error {
+	for _, filterPlugin := range pipeline.Spec.Filters {
+		if err := checkIfPluginIsValid(filterPlugin.Custom, pipeline, pv.deniedFilterPlugins, pipelines); err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
 
-func checkIfPluginIsValid(pluginType, pluginContent string, logPipeline *telemetryv1alpha1.LogPipeline, supportedPlugins, deniedPlugins []string, logPipelines *telemetryv1alpha1.LogPipelineList) error {
-	var err error
-	var section map[string]string
-	var pluginName string
+func (pv *pluginValidator) validateOutputs(pipeline *telemetryv1alpha1.LogPipeline, pipelines *telemetryv1alpha1.LogPipelineList) error {
+	if len(pipeline.Spec.Output.Custom) == 0 {
+		return fmt.Errorf("no output is defined! One output needs to be defined")
```

Comment:
Why does this start with lower case? If it's lower case by purpose, keep the lower case.

```suggestion
		return fmt.Errorf("No output is defined. You must define one output!")
```